ipykernel/ipkernel.py

```diff
@@ -706,6 +724,58 @@ def do_clear(self):
             self.shell.reset(False)
         return dict(status="ok")
 
+    def _associate_identity_of_new_threads_with(self, parent_header):
+        """Intercept the identity of any thread started after this method finished,
+
+        and associate the thread's output with the parent header frame, which allows
+        to direct the outputs to the cell which started the thread.
+
+        This is a no-op if the `self._stdout` and `self._stderr` are not
+        sub-classes of `OutStream`.
+        """
+        stdout = self._stdout
+        stderr = self._stderr
+
+        def start_closure(self: threading.Thread):
+            """Wrap the `threading.Thread.start` to intercept thread identity.
+
+            This is needed because there is no "start" hook yet, but there
+            might be one in the future: https://bugs.python.org/issue14073
+            """
+
+            threading_start(self)
+            for stream in [stdout, stderr]:
+                if isinstance(stream, OutStream):
+                    stream._thread_parents[self.ident] = parent_header
+
+        threading.Thread.start = start_closure  # type:ignore[method-assign]
```

Comment:
I was wondering if there could be a way to not monkey-patch `threading.Thread.start`.
In [akernel](https://github.com/davidbrochart/akernel/blob/71b0b67bb400ff5ce229320662c89f30e02396cd/akernel/kernel.py#L464) it's the `print` function that is monkey-patched, so this PR goes further because it works for lower-level writes to `stdout`/`stderr`, but with monkey-patching there is always a risk that another library changes the same object, and we cannot know that our change will be the last applied.

Reply:
My understanding as per https://bugs.python.org/issue14073 is that there is no other way, but we could mention our use case as another situation motivating introduction of start/exit hooks/callbacks to threads (which is something Python committers have considered in the past but I presume it was not a sufficiently high priority until now).

Reply:
We crossed 'comments', see my other thread on this exact line for (what I think is) a better solution.